### PR TITLE
Console function to force Reset (CU-2e5adgd)

### PR DESCRIPTION
**Pull request by PeterScholtens**

### CODE REVIEW

**Comment by trisapeace** [comment]
Task linked: [CU-2e5adgd Console function to override disable Reset](https://app.clickup.com/t/2e5adgd)

**Comment by PeterScholtens** [comment]
Test1:  Flashed the boron with new firmware, and tried the console function with invalid input such as 0 or any number except for 1, nothing happened except -1 was returned. Entered '1' and received a message before the boron reset. 
Result: Test passed

**Comment by PeterScholtens** [comment]
Test2: Flashed the boron with new firmware and the using the particle console flashed to an old firmware. Like i thought the firmware downloaded but did not reset as it was not pass the DEVICE_RESET_THRESHOLD. Passed a '1' into the force_reset() console function and it reset to the old firmware.
Result: Test passed

**Comment by PeterScholtens** [comment]
Test3: I wrote in the consoleFunctionsTests.cpp
Result: Test passed

File: firmware/README.md
Comment (trisapeace) [inline firmware/README.md:393]: Thanks for noticing this and updating the docs!

File: firmware/boron-ins-fsm/src/consoleFunctions.cpp
Comment (trisapeace) [inline firmware/boron-ins-fsm/src/consoleFunctions.cpp:28]: Any particular reason why this line moved?

File: firmware/boron-ins-fsm/src/consoleFunctions.cpp
Comment (trisapeace) [inline firmware/boron-ins-fsm/src/consoleFunctions.cpp:31]: Oops, looks like you accidentally added some whitespace here. Could you please remove it?

File: firmware/boron-ins-fsm/src/consoleFunctions.cpp
Comment (trisapeace) [inline firmware/boron-ins-fsm/src/consoleFunctions.cpp:289]: Love it.

File: firmware/boron-ins-fsm/src/consoleFunctions.cpp
Comment (trisapeace) [inline firmware/boron-ins-fsm/src/consoleFunctions.cpp:291]: I am personally of the opinion that in C-derived languages, one should put curly braces around the bodies of conditionals and loops. I have two main reasons:
1. If you want to add something to the body of the loop/conditional, then you might forget to add the braces and then your code will be wrong even if the indentation makes it look correct. So difficult to debug
2. If you want to add something to the body of the loop/conditional and you do remember to add the curly braces, then in my preferred style that would add the opening brace on the same line as the if/while/for/etc which would make it show up in the git diff when really nothing there changed.

That being said, I don't change the firmware enough to know what the typical style here is. So I would defer to what is normal in the rest of the code.

File: firmware/boron-ins-fsm/test/mocks/System.h
Comment (trisapeace) [inline firmware/boron-ins-fsm/test/mocks/System.h:13]: It would be really nice if there was some way that we could tell that this function was being called or not. 

Is it possible to add an extern variable in this file for `resetWasCalled`? 
Then
1. Before each test, you can set it to false
2. In this mocked function you will set it to true
3. Then in the test you can check that it is now true, thus showing that the function was, in fact, called

I don't think we're doing this anywhere else yet, but it feels like it should work. Could you please try it out? Let me know if my description isn't clear enough!

Comment (trisapeace) [reply firmware/boron-ins-fsm/test/mocks/System.h:13]: If this works, it would be amazing if you could extend the `mocks/Particle.h` file's `publish` function to do something similar, but instead to put the whole string into the `extern` variable so it can be checked no only that it was called but that it was called with the expected parameters.

File: CHANGELOG.md
Comment (trisapeace) [inline CHANGELOG.md:16]: Thanks for remembering.

Before this is merged, could you please squash your commits and rebase off `main`? Thanks!

**Review by trisapeace — COMMENTED** [review]
Looks great. Just some testing improvements that it would be great if you could do. Thanks!

File: firmware/README.md
Comment (PeterScholtens) [reply firmware/README.md:393]: No worries

File: firmware/boron-ins-fsm/src/consoleFunctions.cpp
Comment (PeterScholtens) [reply firmware/boron-ins-fsm/src/consoleFunctions.cpp]: For sure

File: CHANGELOG.md
Comment (PeterScholtens) [reply CHANGELOG.md:16]: Im not sure how do to that. Usually i rebase and merge.....?

Comment (trisapeace) [reply CHANGELOG.md:16]: The way I normally squash commits together is using the interactive rebase feature in git `git rebase -i HEAD~10` where 10 is the number of commits back in time that you now is more than you're going to need to modify. 

It will display all those 10 (or however many you specify) commits in the text editor with the word `pick` beside each one and a big set of instructions of what words you can change `pick` to in order to do other actions with it. 

Your most recent commit will be at the bottom. If you want to squash it with the commit directly above it in the list, you replace `pick` with `s` (or `squash`).

So in your case, because this whole PR should be a single commit, you can squash all the way back to your first commit (which will still say `pick`). 

Then you save the text file with all the `squash` and `pick` values and it will show you a combined version of the commit message to edit. Edit it down to what you'd like the commit message to be for this one new amazing commit and save it. 

Then if you look at your history using `git log`, you'll only see the one commit with the new message.

This will give us a nice, clean git history when it's committed into `main`. It will also make during the rebase from `main` easier because you will only have to deal with any conflicts once instead of a whole chain of them, once for each commit on your branch.

Like any rebase operation, this changes your git history, so when you next push to GitHub, you'll need to force push (`git push -f`). But don't do this until you're sure that you've got it the way you want it.

If you mess up at any point, try `rebase --abort` or just delete your branch locally and do a fresh `git pull` from GitHub of your branch and try again.

If you'd like to walk through it with me, we can do that too!

File: firmware/boron-ins-fsm/src/consoleFunctions.cpp
Comment (PeterScholtens) [reply firmware/boron-ins-fsm/src/consoleFunctions.cpp:28]: Well I figured that since "Turn_Debugging_Publishes_On_Off" is used the most on the particle console, it would be nicer to just have it at the top of the list. I also inserted the force reset at the top, so it is just under the force update. I can move it back if you do not like it.
Screen Shot 2022-06-27 at 10 34 46 AM

File: firmware/boron-ins-fsm/test/mocks/System.h
Comment (PeterScholtens) [reply firmware/boron-ins-fsm/test/mocks/System.h:13]: It worked for the reset, so ill do it for the publish as well

File: CHANGELOG.md
Comment (PeterScholtens) [reply CHANGELOG.md:16]: Oh ok, cool. Yes if you could walk me through it that would be great! Thank you!

File: firmware/boron-ins-fsm/src/consoleFunctions.cpp
Comment (trisapeace) [reply firmware/boron-ins-fsm/src/consoleFunctions.cpp:28]: Haha, no, that's great. I didn't realize that the order in the code changed the order in Particle. This makes sense to me. Thanks!

**Review by trisapeace — APPROVED** [review]
Looks great! Go ahead and merge once the tests pass.

**Comment by PeterScholtens** [comment]
All tests are passing

**Pull request merged**
